Better support for test modes

mwobensmith opened the pull request:
I have tried to isolate most test logic from the original canary so that it can be refactored into new test modes. To do so, I did the following:

* Removed much code from main.py so that this module only does argument parsing, basic test setup/shutdown and loading of the test mode via new loader module.

* Loader module now is the intermediate between the canary (main.py) and new test modes. It will handle loading of any supported test modes now and in the future.

* Test modes are now in a new subdirectory called "modes" and all have a base class. The base class contains most of the test running code that used to be in main.py. Current test modes are "regression" (equal to what we had before) and "info" (a mode that just generates report-only JSON for the test build). New test modes can inherit from basemode.py.

There is a fair amount of messiness in this code, and many things that are probably not very python-esque. I am completely open to making this cleaner and optimizing where necessary. I believe that the basic approach of a) separating test modes and b) ability to create new ones via inheritance is what we want.

cr commented:
Thanks for that PR! I have some initial feedback. I think that the overall structure is a huge improvement over what we have now. So far I can see a number of code formatting issues that would need to be fixed as well as methods that should be declared @classmethod.

We should discuss what really should be part of the BaseMode class. I think that the way make_profile() works, for example, is rather specific to our SSL regression testing.

But the overall direction is great. We should definitely work on this to get it RTL.

cr commented:
I have a cleanup commit waiting to be pushed to your pull branch once you add me as a collaborator.

A few thoughts:
- What is the `info` mode supposed to do?
- Why does the regular regression scan not write the header any more?
- RegressionMode.__init__() should not use any long-running code. AFAICS it's currently running the whole test suite. Is there a reason for that?

mwobensmith commented:
- What is the `info` mode supposed to do?

This is the scan-only mode. The idea is that it creates a JSON response for all hosts scanned. It can then be used for other tests. I'm thinking that I should re-use it in the upcoming `performance` test mode. 

The name might not be great. I've considered calling it `scan` instead. Thoughts?

- Why does the regular regression scan not write the header any more?

It doesn't? Lines 48-61 of regression.py should do this.

- RegressionMode.__init__() should not use any long-running code. AFAICS it's currently running the whole test suite. Is there a reason for that?

No, no good reason for it. Perhaps I should only use it for initialization stuff, and create a generic start() method in the basemode.py superclass that is guaranteed to be called when the loader wants to start the scan.

cr commented:
My commit addresses several issues. The prevalent ones:

- Lots of indentation issues, especially in modes/regression.py
- Unused imports
- General non-PEP conformant formatting
- Mode classes now save args to self.__args, so no need to pass it to methods.
- ```super(cls, self).method``` is only required to call parent methods that have been overloaded in ```self```. This is commonly the case for ```__init___``` only. For regularly inherited parent methods you can use ```self.method```.

Btw, PyCharm takes good care of formatting issues and imports. Did you give it a try, yet?

mwobensmith commented:
This is totally excellent. I think we should go with it.

Your corrections help me understand python inheritance more. 

I know I should use PyCharm. You did mention it and I intend to check it out, just had not done so.

Should we merge this?

cr commented:
No, this is not RTL, I think. We definitely need to address things like ```__init__``` scope and runtime as well as the mode class interface itself. I am favoring a generic approach of init, setup, run, teardown like everyone knows from tests.

mwobensmith commented:
Totally agree. I will work on implementing those.

cr commented:
Does this make sense for the mode interface?

- ```.__init__(args, ...)``` sets up all state required during object lifetime. It should have the least possible runtime overhead.
- ```.setup()``` prepares everything that is required for (potentially consecutive) test runs. Usually that's everything that takes too long for ```__init__```.
- ```.run()``` does the grunt
- ```.report()``` might make sense as a separate method?
- ```.teardown()``` frees resources allocated by setup

mwobensmith commented:
Perfect sense. Thank you.

mwobensmith commented:
Added standard methods to base test mode class, as well as test modes themselves.

Added new parameter "-x" to determine number of times user wants to scan list. I need to update the regression test to use this new value, since it is currently hard-coded at three passes. 

Added finished performance mode. 

Updated the report function to deal with new values generated by performance mode.

cr commented:
Just added another cleanup commit. Most of the changes were PEP issues and other convention breaks that PyCharm catches right away. Please start using PyCharm.

mwobensmith commented:
This has been superceded by PR #63, closing.
